Utilities/bootstrap

```diff
@@ -694,11 +694,17 @@ def get_swiftpm_env_cmd(args):
             os.path.join(args.build_dirs["swift-crypto"],          "lib"),
             os.path.join(args.build_dirs["swift-system"],          "lib"),
             os.path.join(args.build_dirs["swift-collections"],     "lib"),
-        ] + args.target_info["paths"]["runtimeLibraryPaths"])
+        ]
 
         if platform.system() == 'Darwin':
+            # Need to include backwards compatibility libraries for Concurrency
+            # FIXME: Would be nice if we could get this from `swiftc -print-target-info`
+            back_compat = args.target_info["paths"]["runtimeLibraryPaths"][0] + '../../swift-5.5/macosx'
```

Comment:
does this need to be conditional on the macOS version? iiuc recent Darwin versions do not need this?

Reply:
Yah, that's technically true, but ideally we would be able to make the decision based on the deployment target here, not on the host OS version. @lorentey any opinions?

Reply:
We do not _ever_ want to pick up libswift_Concurrency.dylib from swift-5.5 if the running OS already ships this library in /usr/lib/swift.

There are two paths here, depending on whether the minimum deployment target is below macOS 12 or not.

- If the deployment target is >= macOS 12, then the linker will arrange the executable to load the concurrency lib from `/usr/lib/swift/libswift_Concurrency.dylib`, as that is the OS version that first shipped it.

   ```
   Load command 16
             cmd LC_LOAD_DYLIB
         cmdsize 64
            name /usr/lib/swift/libswiftCore.dylib (offset 24)
      time stamp 2 Wed Dec 31 16:00:02 1969
         current version 5.7.0
   compatibility version 1.0.0
   ```

   In this situation, rdar://75752698 does not apply, so it is possible to use DYLD_LIBRARY_PATH to override the location of this library. I do not suggest we do that, though -- the default configuration will pick up the OS-installed concurrency library, which is the right thing to do, if you want to import Foundation or other high-level SDK components.

- If the deployment target is < macOS 12, then the linker will build the executable so that it picks up the concurrency lib from `@rpath/libswift_Concurrency.dylib`. This will trigger rdar://75752698, so DYLD_LIBRARY_PATH will not work correctly. (Which is not necessarily a big deal, as we do not normally need to use DYLD_LIBRARY_PATH to correctly run such executables.)

   The crucial assumption there is that the executable's RPATH will be configured to have `/usr/lib/swift` followed by the location of the Swift 5.5 backdeployment library. (We need the executable to load the concurrency library from the OS if it's available, and to only load the special back deployment build if the library isn't there.)

   ```
   Load command 16
             cmd LC_LOAD_WEAK_DYLIB
         cmdsize 64
            name @rpath/libswift_Concurrency.dylib (offset 24)
      time stamp 2 Wed Dec 31 16:00:02 1969
         current version 5.7.0
   compatibility version 1.0.0
   Load command 17
             cmd LC_RPATH
         cmdsize 32
            path /usr/lib/swift (offset 12)
   Load command 18
             cmd LC_RPATH
         cmdsize ??
            path <path to the 5.5 deployment libs directory>
   ```

   For this to work properly, the RPATH needs to be explicitly configured while building the executable -- the compiler has to be told where the back deployment libraries are going to be at runtime: it cannot on its own figure this out. (In app bundles, the correct path is an executable-relative path within the app bundle (as Xcode copies the library there. For toolchain components, the right path is an executable-relative path to the swift-5.5/macosx directory within the toolchain. For executables that are not part of the toolchain (such as ones built by SwiftPM, or ones built manually but not intended to be installed on the toolchain), the right option is probably to use an absolute path to the swift-5.5/macosx directory within the specific toolchain that was used to build the executable.)

So I recommend _not_ configuring a DYLD_LIBRARY_PATH in either of these cases.

If you're able to bump the minimum deployment target to macOS 12, then the easiest thing is to do that. (This makes it technically possible to use DYLD_LIBRARY_PATH to pick up a custom dev build of the stdlib at runtime, but I recommend against doing that, at least by default.)

If bumping the deployment target is not an option, then I recommend building executables with `-Xlinker -rpath -Xlinker <path>` where `<path>` is the (`@executable_path`-relative or absolute) path to the toolchain's swift-5.5/macosx directory. To verify things are as expected, compare `otool -l` output with the example above: (1) the load command for libswift_Concurrency needs to be `@rpath`-relative, and (2) the LC_RPATH command for /usr/lib/swift needs to precede the LC_RPATH load command for the swift-5.5 directory.

(Note: Do not use the `-toolchain-stdlib-rpath` compiler option outside of the Swift compiler test suite. It generates binaries that are hardwired not to load stdlib components from /usr/lib/swift, which isn't going to work, especially if the binary loads any SDK components above the stdlib.)

Reply:
Thanks for the detailed explanation! I think switching to `-rpath` is the right option here.